flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/CallContext.java

```diff
@@ -93,6 +93,17 @@ default ValidationException newValidationError(String message, Object... args) {
         return new ValidationException(String.format(message, args));
     }
 
+    /**
+     * Shorthand for {@code if (throwOnFailure) throw ValidationException(...) else return
+     * Optional.empty()}.
+     */
+    default <T> Optional<T> failInference(boolean throwOnFailure, String message, Object... args) {
```

Comment:
TBH I think `fail` is generally more meaningful. `emptyOrError` gives me the idea that whether to return empty or error is a decision of the function, but it's not really the case since here is only `throwOnFailure` which decides whether or not to throw the exception. Also `failInference` reads nicer in the code. Maybe just `fail` could be ever better?